3 feature(View): Выполнена лабораторная работа №3 по ООАиПр

Melaved opened the pull request.

Melaved committed:
feature(View): сделан макет верстки для 3 лабораторной

Melaved committed:
feature(View): Релиазованы команды
- добавлен класс RelayCommand
- в классе MainVM реализованы команды(пока что не работают)
- реализовал конвертер значений BooleanToVisibilityConverter для сокрытия кнопки Apply
-  выполнена привязка кнопок и конвертера значений к форме

Melaved committed:
feature(View): Переделана сериализация и исправлены команды

- Исправлена работа команд чтобы все соответствовало методичке
- класс ContactSerializer теперь сохраняет данные между сессиями
- удалены классы SaveCommand и LoadCommand
- Поправлена верстка
- Требуется рефакторинг

Melaved committed:
refactor(View): Добавлены комментарии

- добавлены комментарии
- изменен порядок содержимого классов согласно стандарту оформления кода

inittose commented on View/Converters/BooleanToVisibilityConverter.cs:
```diff
@@ -0,0 +1,48 @@
+{
+    /// <summary>
+    /// Конвертер, преобразующий значение типа <see cref="bool"/> в значение перечисления <see cref="Visibility"/>.
+    /// </summary>
```
Комментарии, как и код, не должены превышать 100 символов в длину
Проверить во всем решении, в т.ч. код

inittose commented on View/Converters/BooleanToVisibilityConverter.cs:
```diff
@@ -0,0 +1,48 @@
+            if (value is bool boolValue)
+            {
+                return boolValue ? Visibility.Visible : Visibility.Collapsed;
+            }
+            return Visibility.Collapsed;
```
Добавить пустую строчку. После блока конструкции if должна идти разделяющая пустая строка
Проверить во всем решении

inittose commented on View/MainWindow.xaml.cs:
```diff
@@ -16,5 +16,21 @@ public MainWindow()
             DataContext = new MainVM();
         }
+
+        private void Button_Click(object sender, RoutedEventArgs e)
+        {
+            var nameExpression = NameTextBox.GetBindingExpression(TextBox.TextProperty);
+            var phoneExpression = PhoneNumberTextBox.GetBindingExpression(TextBox.TextProperty);
+            var emailExpression = EmailTextBox.GetBindingExpression(TextBox.TextProperty);
+
+            nameExpression?.UpdateSource();
+            phoneExpression?.UpdateSource();
+            emailExpression?.UpdateSource();
+
+            if (DataContext is MainVM ViewModel)
+            {
+                ViewModel.ApplyContact(null);
+            }
+        }
```
Уйти от использования code behind событий
Вся логика должна реализовываться в части ViewModel

inittose commented on View/Model/Contact.cs:
```diff
@@ -19,29 +36,66 @@ public Contact(string name, string phoneNumber, string email)
+        public event PropertyChangedEventHandler PropertyChanged;
 
         /// <summary>
-        /// Возвращает и задает телефонный номер контакта.
+        /// Получает или задает имя контакта.
         /// </summary>
```
"Получает или задает" заменить на "Возвращает и задает"
Проверить во всем решении

inittose commented on View/Model/Services/ContactSerializer.cs:
```diff
@@ -1,62 +1,42 @@
         /// </summary>
-        /// <returns>Объект <see cref="Contact"/>, если файл существует, иначе null.</returns>
-        public static Contact LoadContact()
+        /// <returns>
+        /// Возвращает список контактов, если файл существует и успешно десериализован.
+        /// В противном случае возвращает пустой список.
+        public static List<Contact> LoadContacts()
```
Нет закрывающего тэга \</returns\>

inittose commented on View/Model/Services/ContactSerializer.cs:
```diff
@@ -1,62 +1,42 @@
-        /// Путь к файлу, в котором хранятся контакты.
-        /// </summary>
-        public static string FilePath { get; }
+        private static readonly string FilePath = "contacts.json";
 
```
Файлы сохранения/конфигурации/настроек программы должны находиться по пути:
`%AppData%\Contacts`
Где AppData - переменная среды (`C:\Users\<ИМЯ_ПОЛЬЗОВАТЕЛЯ>\AppData\Roaming`)
Если ты задаешь поле (а не свойство), то оно должно называться `_filePath`

inittose replied:
Если возникают вопросы, задавай их под замечанием или в чате телеграм

inittose replied:
contacts.json все еще находится внутри папки debug (рядом с собранным приложением), а должен находится в AppData

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -1,121 +1,274 @@
+using View.Model.Services;
 using View.Model;
+using View.ViewModel;
+using System.Windows.Data;
+using System.Windows.Controls;
```
Не используются

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -1,121 +1,274 @@
+
+    /// <summary>
+    /// Коллекция контактов, отображаемых в главном окне.
+    /// </summary>
+    public ObservableCollection<Contact> Contacts { get; set; } = new ObservableCollection<Contact>();
+
```
Строка больше 100 символов в длину, либо перевести на новую строку, либо упростить до:
```csharp
public ObservableCollection<Contact> Contacts { get; set; } = [];
```

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -1,121 +1,274 @@
-            set
+            if (_isAddingNewContact && value != null)
+            {
+                _isAddingNewContact = false;
+                IsApplyButtonVisible = false;
+                IsEditMode = false;
+                OnPropertyChanged(nameof(IsAddingNewContact));
+            }
+
+            if (_isEditingContact && value != null)
             {
-                _contact = value;
-                OnPropertyChanged(nameof(Contact));
+                _isEditingContact = false;
+                IsApplyButtonVisible = false;
+                IsEditMode = false;
+                OnPropertyChanged(nameof(IsEditingContact));
             }
```
Разве IsAddingNewContact и IsEditingContact не выполняют одну и ту же функцию?

inittose commented on View/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,57 @@
+        /// </summary>
+        public event EventHandler CanExecuteChanged
+        {
+            add { CommandManager.RequerySuggested += value; }
+            remove { CommandManager.RequerySuggested -= value; }
```
Заменить на:
```csharp
add => CommandManager.RequerySuggested += value;
remove => CommandManager.RequerySuggested -= value;
```

inittose commented on View/MainWindow.xaml:
```diff
@@ -2,120 +2,207 @@
+    xmlns:model="clr-namespace:View.Model"
+    xmlns:viewmodel="clr-namespace:View.ViewModel"
     Title="Contacts"
     Width="525"
     Height="375"
```
Добавить сюда минимальную возможную ширину и высоту окна

inittose replied:
Сделать минимальную возможную ширину побольше, чтобы не было обрезания текста у кнопки Remove
<img width="174" alt="image" src="https://github.com/user-attachments/assets/2ce25e5b-a883-43af-8118-c60f67c4d8de" />

inittose commented:
Когда поправишь замечания, нажми на кнопку Re-request review
![изображение](https://github.com/user-attachments/assets/8263758a-4e34-4c7e-9c06-5dc609fdc677)

Melaved committed:
feature(View): Исправлены все замечания

- убрано использование code behind
- исправлены комментарии
- поправлены все строки
- убран второй флаг(режим)
- добавлена минимальная ширина и высота окна

inittose commented on View/Converters/BooleanToVisibilityConverter.cs:
```diff
@@ -5,42 +5,51 @@
         /// </returns>
-        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
+        public object Convert
+            (object value, Type targetType, object parameter, CultureInfo culture)
```
Когда метод длинее 100 символов его разбивают на новую строку таким образом:
```csharp
        public object Convert(
            object value,
            Type targetType,
            object parameter,
            CultureInfo culture)
```
Т.е. каждый аргумент на своей строчке
Проверить во всем решении

inittose commented on View/Converters/BooleanToVisibilityConverter.cs:
```diff
@@ -5,42 +5,51 @@
 {
     /// <summary>
-    /// Конвертер, преобразующий значение типа <see cref="bool"/> в значение перечисления <see cref="Visibility"/>.
+    /// Конвертер, преобразующий значение типа <see cref="bool"/> 
```
Лишний пробел в конце строки комментария
Проверить во всем решении
Проверить на пробелы может помочь StyleCop, файл конфигурации можно найти на ЯДиске:
`50ohm_Students/ГПО/Статьи по процессу разработки (копии с Wiki 50ohm)/StyleCop/.editorconfig`

inittose replied:
Я ошибся, StyleCop не показывает есть ли лишние пробелы, это показывает расширение Resharper для VisualStudio
Можешь поставить его или просто сочетанием клавиш включить визиализацию пробелов: `Ctrl + R, Ctrl + W`

inittose commented on View/Model/Services/ContactSerializer.cs:
```diff
@@ -28,14 +28,15 @@ public static void SaveContacts(IEnumerable<Contact> contacts)
         /// <returns>
         /// Возвращает список контактов, если файл существует и успешно десериализован.
-        /// В противном случае возвращает пустой список.
+        /// В противном случае возвращает пустой список.</returns>
```
\</returns\> с новой строки

inittose commented on View/Model/Services/ContactSerializer.cs:
```diff
@@ -1,62 +1,43 @@
 ﻿using System.IO;
-using Newtonsoft.Json;
+using System.Text.Json;
```
ContactSerializer реализуй с помошью Newtonsoft.Json, либо удали его из nuget-пакетов

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -1,121 +1,253 @@
+                }
+
+                IsApplyButtonVisible = false;
+                IsEditMode = false;           
```
Лишние пробелы в конце строки

inittose replied:
В конце 210 строки все еще лишние пробелы

Melaved replied:
исправил новый коммит, у меня почему то когда коммит отправлял файл MainVM отменился

Melaved replied:
исправил замечания добавил новый коммит*

inittose commented on View/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace View.ViewModel
+{
+    public class RelayCommand : ICommand
```
Нет комментария к классу

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -1,121 +1,253 @@
+    /// <param name="parameter">Параметр команды.</param>
+    public void RemoveContact(object parameter)
+    {
+        if (SelectedContact != null)
```
Инвертировать if, чтобы убрать вложенность, чтобы повысить читаемость кода
Это можно сделать так:
```csharp
if (SelectedContact == null)
{
    return;
}
```
Посмотреть, где еще можно убрать лишнюю вложенность

inittose commented on View/MainWindow.xaml.cs:
```diff
@@ -1,5 +1,5 @@
 ﻿using System.Windows;
-using View.ViewModel;
+using System.Windows.Controls;
```
Не используется

Melaved committed:
feature(View): Исправлены замечания

- изменен класс ContactSerializer, теперь он использует Newtonsoft.Json
- убрана лишняя вложенность
- отформатированы строки

Melaved committed:
style(View):  Исправил замечания

inittose commented on View/Converters/BooleanToVisibilityConverter.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <summary>
+        /// Преобразует значение обратно из <see cref="Visibility"/> в <see cref="bool"/>.
+        /// </summary>
+        /// <param name="value">Значение типа <see cref="Visibility"/>, 
```
Лишний пробел в конце строки

inittose commented on View/Converters/BooleanToVisibilityConverter.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <exception cref="NotImplementedException">Метод не реализован.</exception>
+        public object ConvertBack(
+            object value,
+            Type targetType, 
+            object parameter, 
```
Лишний пробел в конце строки

inittose commented on View/Converters/InverseBooleanConverter.cs:
```diff
@@ -0,0 +1,65 @@
+        /// В противном случае возвращает исходное значение <paramref name="value"/>.
+        /// </returns>
+        public object Convert(
+            object value, 
```
Лишний пробел в конце строки

inittose commented on View/Converters/InverseBooleanConverter.cs:
```diff
@@ -0,0 +1,65 @@
+        /// </returns>
+        public object ConvertBack(
+            object value,
+            Type targetType, 
```
Лишний пробел в конце строки

inittose commented on View/MainWindow.xaml.cs:
```diff
@@ -16,5 +15,6 @@ public MainWindow()
             InitializeComponent();
             DataContext = new MainVM();
         }
+
```
Лишняя пустая строка

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -1,121 +1,259 @@
-        /// Свойство для доступа к номеру телефона контакта.
-        /// </summary>
-        public string PhoneNumber
+        int index = Contacts.IndexOf(SelectedContact);
```
Используй неявное типизирование локальных переменный
int замени на var

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -1,121 +1,259 @@
+/// <summary>
+/// Представляет ViewModel для главного окна приложения.
+/// </summary>
+public class MainVM : INotifyPropertyChanged
```
У класса отсутствует namespace

Melaved committed:
style(View):

Melaved committed:
style(View): Исправил замечания
- установил и настроил ReSharper
- убрал лишние пробелы

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -1,121 +1,264 @@
+using System.Windows.Input;
 using View.Model;
+using View.Model.Services;
+using View.ViewModel;
```
using View.ViewModel не используется

Melaved committed:
feature(View): Исправил ContactSerializer
